Limit places db queries to prevent critical errors

I updated to the latest activity stream on dev, and found that my highlights were not updating.  When I started digging into browser consoles and toolboxes, I found that it was caught in an infinite loop of re requesting metadata, failing, and trying again.  

When I dug a little further, I found that the highlights feed refresh logic was waking up, querying places db for urls, finding 1014 candidate urls, passing those to preview provider, which breaks them into lots of 20 to send to the metadata service, makes 51 concurrent requests to metadata service, attempts to insert 1014 metadata entries into metadatastore, catastrophically fails, and then tries to do the whole thing over again.  

We should limit the number of URLs we receive from placesdb and attempt to send to preview provider.

I've logged some things here to help with debugging, but this caused my activity stream to become unresponsive and I was forced to uninstall.  For reference, I have both local metadata parsing and metadata service enabled.

![screenshot 2016-11-09 15 17 48](https://cloud.githubusercontent.com/assets/119884/20154031/88a0d8ae-a693-11e6-9ac1-5cb2acb0e5c1.png)

Looks like there is an ESLint rule for this (of course), [`no-template-curly-in-string`](http://eslint.org/docs/rules/no-template-curly-in-string):

```ini
"no-template-curly-in-string": "error",
```

It caught 2-3 of the issues in /addon/MetadataStore.js (but not the third, since it was missing the `$`). But also finds a few false positives in /content-test/, which I'll have to disable errors for:

```js
$ ./node_modules/.bin/eslint .

/Users/pdehaan/dev/github/mozilla/activity-streams/addon/MetadataStore.js
  448:22  error  Unexpected template string expression  no-template-curly-in-string
  471:24  error  Unexpected template string expression  no-template-curly-in-string

/Users/pdehaan/dev/github/mozilla/activity-streams/content-test/lib/getRelativeTime.test.js
   9:6  error  Unexpected template string expression  no-template-curly-in-string
  13:6  error  Unexpected template string expression  no-template-curly-in-string
  17:6  error  Unexpected template string expression  no-template-curly-in-string
  21:6  error  Unexpected template string expression  no-template-curly-in-string

✖ 6 problems (6 errors, 0 warnings)
```

--- a/addon/MetadataStore.js
+++ b/addon/MetadataStore.js
@@ -86,6 +86,9 @@ const IMAGE_TYPES = {
   "preview": 3
 };
 
+/* Maxium number of allowed parameters in a single SQL statement (defined here: https://www.sqlite.org/limits.html) */
+const SQL_PARAMETER_LIMIT = 999;
+
 function MetadataStore(path, migrations = null) {
   this._path = path || OS.Path.join(OS.Constants.Path.profileDir, METASTORE_NAME);
   this._migrations = migrations || MIGRATIONS;
@@ -433,13 +436,17 @@ MetadataStore.prototype = {
    * Returns a promise with the array of retrieved metadata records
    */
   asyncGetMetadataByCacheKey: Task.async(function*(cacheKeys) {
-    const quoted = cacheKeys.map(key => "?").join(",");
+    let limitedCacheKeys = cacheKeys;
+    if (limitedCacheKeys.length > SQL_PARAMETER_LIMIT) {
+      limitedCacheKeys = cacheKeys.slice(0, SQL_PARAMETER_LIMIT);
+    }
+    const quoted = limitedCacheKeys.map(key => "?").join(",");
     let metaObjects;
     try {
       metaObjects = yield this.asyncExecuteQuery(
         `SELECT * FROM page_metadata WHERE cache_key IN (${quoted})`,
         {
-          params: cacheKeys,
+          params: limitedCacheKeys,
           columns: ["id", "cache_key", "places_url", "title", "type", "description", "media_url", "provider_name"]
         }
       );

